bin/propolis-server/src/lib/vm/request_queue.rs

```diff
@@ -164,248 +214,330 @@ enum RequestDisposition {
     Deny(RequestDeniedReason),
 }
 
-/// The current disposition for each kind of incoming request.
+/// A kind of request that can be popped from the queue and then completed.
 #[derive(Copy, Clone, Debug)]
-struct AllowedRequests {
-    start: RequestDisposition,
-    migrate_as_source: RequestDisposition,
-    reboot: RequestDisposition,
-    mutate: RequestDisposition,
-    stop: RequestDisposition,
+pub(super) enum CompletedRequest {
+    Start { succeeded: bool },
+    Reboot,
+    MigrationOut { succeeded: bool },
+    Stop,
+}
+
+/// The queue's internal notion of the VM's runtime state.
+#[derive(Copy, Clone, Debug)]
+enum QueueState {
+    /// The instance has not started yet and no one has asked it to start.
+    NotStarted,
+
+    /// The instance is not running yet, but the state driver will eventually
+    /// try to start it.
+    StartPending,
+
+    /// The instance has successfully started.
+    Running,
+
+    /// The instance has stopped due to a migration out.
+    MigratedOut,
+
+    /// The instance has shut down.
+    Stopped,
+
+    /// The instance failed to start.
+    Failed,
+}
+
+impl QueueState {
+    /// If `self` is a state in which new change requests should be denied
+    /// unconditionally, returns a `Some` containing an appropriate
+    /// [`RequestDeniedReason`]; returns `None` otherwise.
+    fn deny_reason(&self) -> Option<RequestDeniedReason> {
+        match self {
+            Self::MigratedOut => Some(RequestDeniedReason::MigratedOut),
+            Self::Stopped => Some(RequestDeniedReason::Halted),
+            Self::Failed => Some(RequestDeniedReason::InstanceFailed),
+            _ => None,
+        }
+    }
 }
 
-/// A queue for external requests to change an instance's state.
 #[derive(Debug)]
-pub struct ExternalRequestQueue {
-    queue: VecDeque<ExternalRequest>,
-    allowed: AllowedRequests,
+pub(super) struct ExternalRequestQueue {
+    /// The queue of unprocessed state change requests.
+    state_queue: VecDeque<StateChangeRequest>,
+
+    /// The queue of unprocessed component change requests.
+    component_queue: VecDeque<ComponentChangeRequest>,
```

Comment:
i note that this means that there's no longer a total ordering of state change and component change requests, but i think that...may not matter.